Refactor priorities system

pastk opened the pull request:
This refactoring is a biggest part of #4314.

Biggest changes:
- remove the minVisibleScale overlays bucketing (the biggest pain in the old system)
- remove the fixed z10 shields exception
- remove exceptional handling of pathtexts priorities
- disable usage of DisplayFlag in the displacement logic
- remove `-x-me-*` priority hacks from style files
- make overlays and geometry priorities independent
- remove z-index declarations from style files
- put all priorities into auto-formatted and sorted prio.txt files (gives a good overview of the priorities order)
- a separate prio.txt file per each priority range (overlays, foreground, etc) makes it explicit and obvious what is rendered above what
- make all optional captions below other overlays
- re-arrange priorities into a more logical order and fix many inconsistencies and bugs

(there will be more PRs to remove some more minor exceptions and cleanup/refactor, improve sub-styling, etc.)

A corresponding kothic change: https://github.com/organicmaps/kothic/pull/10

Closes:
- #4426 
- #4428 
- #4590 
- #4506 and #4538 


Following is the original first post:
-------

As per the plan https://github.com/organicmaps/organicmaps/issues/4314#issuecomment-1442367661
The final goal will be to have a single easily editable priorities file without "zoom buckets" restrictions and with geometry priorities detached from overlays ones.
	
This is a first step - moving minVisibleScale logic to kothic, so that style priorities will incorporate current minVisibleScale ordering.
	
All priority numbers changed here as they include minVisibleScale component now, but relative types ordering didn't change.
		
However there are behaviour changes still as some of minVisibleScale logic was run-time specific:
- run-time `feature::GetMinDrawableScale()` result was dependent on area size, i.e. if an area was deemed to small to render at some zoom levels, then resulting minVisibleScale value was higher than set in styles (actually it would be good to give big size area POIs some more priority, but in a more subtle way than this);
- run-time minVisibleScale was calculated per feature, i.e. min value across all of its types (causes #4590), while in kothic its per type; **TODO:** a proper fix needs to be done in #4590 first;
- path texts were hardcoded not to use minVisibleScale for prioritization at all (#4428), now they use it; **TODO:** adjust path texts priorities so that they're below important POIs like bus stops and train stations (will be possible when geometry and overlays priorities are detached).

biodranik commented:
What are the pros and cons of this approach?

pastk commented:
> What are the pros and cons of this approach?

The general "zoom bucketing" pros and cons were discussed here https://github.com/organicmaps/organicmaps/issues/4314#issuecomment-1425992776

I'd just add that since then I saw many cases when z-indexes were changed without awareness of this mechanism - hence in general it leads to a lot of confusion and side-effects/bugs. So now I'm more confident it'll be much better without enforced "zoom bucketing" (but minVisibleScale could be added to the comments of the priorities file to give some guidance to the designers, as in general minVisibleScale should correlate with z-indexes still IMO).

pastk commented:
Added a commit to calculate minVisibleScale values for overlays only (kothic change: https://github.com/organicmaps/kothic/pull/10/commits/6298d94c34b2120474edd2920a4caa6e9ec132a7).

Why?
ATM minVisibleScale is calculated as the minimum value for all draw elements a type may have.
This means that e.g. a `highway-motorway` path text is in z7 zoom bucket, although its visible from z10 only - this is because motorway's line is being drawn from z7. Or e.g. `natural-water-lake` captions have minVisibleScale of z1, although they're visible from z10 only - because big lakes areas are rendered from z1.
Hence overlays (icons, captions) of types which have geometry that is rendered early (i.e. most roads/lines and area POIs) have their overlays priorities boosted significantly and displace node POIs around (although they might have a higher z-index). E.g. I think this was the reason for z10 minVisibleScale hardcode for highway shields (#4426) - to make shield's priorities lower.
Its more logical if motorway path texts (and shields) and lake's captions are in z10 zoom bucket - just because they appear at this zoom level. Does it make sense?

So this change makes minVisibleScale calculated for overlays only (icons, captions, path texts, shields), i.e. it doesn't include visibility of lines and area fills (and they never used minVisibleScale logic anyway).
Please note that minVisibleScale is calculated as the min value for all overlays, e.g. if type's icon is visible from z14 and its caption from z16, then both of them will be in z14 zoom bucket. Its possible to make the caption to be in z16 bucket in this case, but it will make priority distance between captions and icons much bigger than the current `1000` priority gap (https://github.com/organicmaps/organicmaps/issues/4314#issuecomment-1429379164), so it might make sense only if we for sure choose to have independent icons and captions priorities - which I'm skeptical of.

So this change effectively lowers too high priorities of overlays of many lines and areas:
- it removes z10 minVisibleScale hardcode for highway shields (#4426) as its no longer necessary;
- it lowers priorities of highways' path texts, hence helping with #4428, but not enough still to solve street names displacing e.g. bus stops issue (because bus stops are in a very low z16 zoom bucket) - **to be resolved later**
- it helps partly with #4590 too by reducing `building`'s captions priority from z14 to z16 zoom bucket.

I.e. I think the most of the priority changes are positive and make sense, however there might be cases when we want to retain higher overlay priorities?
E.g. maybe we don't want isolines' captions to be any lower priority (related to #4537)?
Are there any other changes that look negative? I can tackle them in subsequent commits. Please take a look at the ordering changes in https://github.com/organicmaps/organicmaps/pull/4645/commits/13247ed46a20616ddde26d52ae0897829ebc9f5c
(e.g `OVRL/z14 !z12   text   natural-bare_rock   z[14-19]` means that the caption was in z12 zoom bucket initially and this change moved it to z14).

biodranik commented:
It would be great to compare equal videos with zooms/scrolls over some areas with different patches, in parallel. Is there a tool that we can use? Revive `mapshot.cpp` ?

pastk commented:
I'm thinking of making a few branches showcasing different approaches indeed, but have no idea how to run them "in parallel".

pastk commented:
Dumped priority values into new separate mapcss files (see them in https://github.com/organicmaps/organicmaps/pull/4645/commits/e4d1e207a4291fe52976d563b55fbdd23959f1f3).

I think its better to have them separated into a few priority range specific files (e.g. overlays, fg, bg-by-size..) instead of lumping together into a single file (as it gives a false impression of a single priorities space).

The files use the existing mapcss syntax atm. Later we can change it to make the files more readable.

The next step is to read these files and use priorities directly from them instead of from existing z-index declarations.

pastk commented:
> The files use the existing mapcss syntax atm. Later we can change it to make the files more readable.

I didn't consider that inheritance between types will get into way in a prioritues-sorted file.
E.g.
```
[aeroway=aerodrome][international] {priority: 150}
[aeroway=aerodrome] {priority: 100}
```
`[aeroway=aerodrome][international]` will get overriden with `priority: 100`.

It was possible to patch kothic to disable inheritance just for priorities, but its very hacky.
Also needed to fight with kothic type-matching bugs (`[access=no]` doesn't match `[access=no]`!). And a lot of newly added (one per each priority) style choosers / selectors made drules generation 3x times slower..

So I decided its better to go a new simple priorities files syntax from the start: https://github.com/organicmaps/organicmaps/pull/4645/commits/fc1bc1aeaa4bb8ff3db3873f0519e40e9a89ee68

The only downside is that they use type names directly already, while other style files use tags-like matching of mapcss-mapping.csv rules still.

Also I included into this PR all the priority fixes from #4506 and #4538 as those exceptions require setting zoom-specific priorities.

> 
> The next step is to read these files and use priorities directly from them instead of from existing z-index declarations.

This is done.
Priorities are loaded from the new files. The files are re-formatted automatically and there is basic validation.

Also I compressed / re-spaced all priorities (with step 10) to make them look nicer and easier editable.

The syntax could be improved later, some ideas:
- group drules with identical priorities
- display first zoom level when a drule appears - not editable, for designers' information only
- preserve custom comments
- improve sub-styling / inheritance capabilities

Next I will remove old z-index declarations, priority hacks etc. from the mapcss files.

pastk commented:
@vng @biodranik it would be great to have some feedback at this point!

biodranik commented:
Faster drules generation is important for "real-time" style editor implementation.

biodranik commented on data/visibility.txt:
```diff
@@ -533,11 +533,11 @@ world  00000000000000000000  +
             sculpture  00000000000000011111  -
             statue  00000000000000011111  -
-            stolperstein  00000000000000011111  -
+            stolperstein  00000000000000000000  -
```
?

pastk replied:
Those types were "displayed" using a zero-icon in vehicle style (so were invisible actually). Now they're removed from there properly.

biodranik commented on data/visibility.txt:
```diff
@@ -533,11 +533,11 @@ world  00000000000000000000  +
-            cross  00000000000000011111  -
-            plaque  00000000000000011111  -
+            cross  00000000000000000111  -
+            plaque  00000000000000000000  -
```
?

biodranik commented on drape/overlay_handle.cpp:
```diff
@@ -232,7 +232,7 @@ std::string SquareHandle::GetOverlayDebugInfo()
 uint64_t CalculateOverlayPriority(int minZoomLevel, uint8_t rank, float depth)
 {
-  ASSERT(0 <= depth && depth <= 100000, (depth));
+  ASSERT(0 <= depth && depth <= 190000, (depth));
```
A comment why 190K would be helpful.

pastk replied:
this function will need to be cleaned/refactored later to remove remnants of minVisibleScale anyway...

pastk replied:
it was actually a temporary rise for migration to work (some overlays had priorities as high as 182k);
now after priorities compression and re-spacing we can put back 100000 or even 10000 is more than enough

biodranik commented on drape_frontend/apply_feature_functors.cpp:
```diff
@@ -1015,7 +1018,7 @@ void ApplyLineFeatureAdditional::GetRoadShieldsViewParams(ref_ptr<dp::TextureMan
     poiParams.m_depthTestEnabled = false;
     poiParams.m_depthLayer = DepthLayer::OverlayLayer;
-    poiParams.m_minVisibleScale = kShieldMinVisibleZoomLevel;
+    poiParams.m_minVisibleScale = m_minVisibleScale;
```
Can some commits/fixes be reviewed/merged separately from this PR?

biodranik reviewed (commented):
Can we "eat an elephant one bite at a time"?

pastk commented:
> Can we "eat an elephant one bite at a time"?

Yeap I really want to limit the scope of this PR, but things are very intertwined here at the moment.

The only things that could be reviewed and merged separately are priority fixes
https://github.com/organicmaps/organicmaps/pull/4506 and https://github.com/organicmaps/organicmaps/pull/4538

pastk commented:
> Next I will remove old z-index declarations, priority hacks etc. from the mapcss files.

Done!

Next I'll see if some priorities needs amendment due to removal of hard-coded exceptions, e.g. #4428

pastk commented:
@vng @biodranik @RedAuburn @dvdmrtnz and anyone else interested :)

Need your feedback for the following before/after screenshots. Anything to fix / re-prioritize?

Even better - build this PR/branch for yourself and browse some area you know very well and give feedback on POIs priorities.
(only main style, I'll update the vehicle one later)

![drop-mvs-z4](https://github.com/organicmaps/organicmaps/assets/18434508/c52f403c-ac73-46fb-99f1-90417b3fcfde)

![drop-mvs-z11](https://github.com/organicmaps/organicmaps/assets/18434508/22cbc182-d7e6-4492-8dd0-63e53128b5a7)

![drop-mvs-z13](https://github.com/organicmaps/organicmaps/assets/18434508/9c3f3f2c-9577-4f02-90ad-fed5908c4644)

![drop-mvs-z14](https://github.com/organicmaps/organicmaps/assets/18434508/fb899648-2675-4708-916d-0f2500b72850)

![drop-mvs-z16](https://github.com/organicmaps/organicmaps/assets/18434508/6a53e42a-d6ee-485c-9499-16ce28827b37)

vng commented:
4 scale - prioritize cities over seas, good
11 scale - prioritize bold and outline city name, good
13 scale - prioritize railway stations, good
16 scale - bus stops?

pastk commented:
> 14 scale - didn't get

- less small streams names
- more neighborhoods, forests/parks

btw, I'm not sure what should get more priority?
`landuse-forest` or `leisure-park`?

> 16 scale - bus stops?

yes
and many minor changes

atm I made `amenity-parking` icons higher than `highway-tertiary` pathtexts but lower than secondary

pastk commented:
> Done! Started with pathtext priorities but then couldn't stop and adjusted a lot of things, for details see [8213b60](https://github.com/organicmaps/organicmaps/commit/8213b60728db57a41c346df88ecac2de0b00ee26)
> 
> The ability to overview the real priority order makes apparent that many things were wrong before :)

So it was a first experience adjusting priorities using a new system.

Overall is great, much more obvious and easier than before, no hidden hardcoded nuances and exceptions, ability to adjust geometry and overlays priorities separately, etc.

There is room for improvement still.

1. It'll be even greater if we could just change only the order of lines. Now after re-shuffling the lines one has to manually adjust priority numbers accordingly.. But then we need to find a way to support following cases w/o using prio numbers:
  - sub-styling (a child style should contain changes only, not a copy of the whole priorities file)
  - there is an idea how to support proper rendering of landuse areas located on bridges and rooftops - it requires numbers also...
We've discussed it in #4314, but there is no elaborated number-less approaches yet. One new raw idea is to keep the numbers, but somehow automate re-numbering of the moved lines only.

2. Its very cumbersome to adjust priorities of both icons and captions (they're independent, same for shields + pathtexts):
  - separate icon and captions make overlays priorities file twice as big, hence harder to overview and maintain
  - its very easy to adjust icons and then forget to do the same for captions...
There is an idea to make captions the same priority as icons always, it was discussed before. The biggest concern with it is that captions might displace too many icons around. I'll do some real screenshots comparisons.

biodranik commented:
For zoom 11, forest icons doesn't look useful. Also, ship stops without titles are not so useful too (town names are more useful).

biodranik commented:
Can you please force-push this updated branch to the beta branch? We'll build apks and iOS versions to test.

pastk commented:
> For zoom 11, forest icons doesn't look useful. Also, ship stops without titles are not so useful too (town names are more useful).

There are much less ferry terminal icons after this PR, so all good here?

(Named) forest icons start from z12.
Those z11 icons are nature_reserves and protected_areas.
Now they appeared in place of de-prioritized road shields, stream names, ferry terminals and waterfalls. Do you think reserve's priority should be lowered?

vng commented:
Have doubts about parking vs tertiary names. Street names in general are a pain for now cause they are displaced by POI icons like bicycle parking and regular parking (which are usually placed near a street). And it is often hard to find a street name.

pastk commented:
> Have doubts about parking vs tertiary names. Street names in general are a pain for now cause they are displaced by POI icons like bicycle parking and regular parking (which are usually placed near a street). And it is often hard to find a street name.

Only regular parkings icons are higher than tertiary.
Bicycle and motorcycle parkings are lower than any road pathtexts in this PR.

pastk commented:
Main style, made
- regular parkings above highway-residential only
- nature_reserves below motorways/trunk roads
- hospitals/universities higher than leisure-park
- forests below than leisure-park

Vehicle - many changes similar to main, and some specific
- highways on top
- higher priority of highway-services, rest_areas, car_repair, car_wash, toll_booth, fuel/charging stations etc.

There are still priority exceptions hardcoded in the drape for the navigation mode to make pathtexts and shields higher than any other overlays.
This affects **all** pathtexts incl. e.g. rivers :) so they displace everything around. Also all highways pathtexts are treated the same, so e.g. residential road name can displace trunk road one. We'll get rid of these exceptions later.

![veh_z12](https://github.com/organicmaps/organicmaps/assets/18434508/81e6945f-bec5-4df8-b8ed-74959a8382a7)

pastk commented:
> 3\. Its very cumbersome to adjust priorities of both icons and captions (they're independent, same for shields + pathtexts):
> 
> 
>     * separate icon and captions make overlays priorities file twice as big, hence harder to overview and maintain
> 
>     * its very easy to adjust icons and then forget to do the same for captions...
>       There is an idea to make captions the same priority as icons always, it was discussed before. The biggest concern with it is that captions might displace too many icons around. I'll do some real screenshots comparisons.

Here is real comparisons (2nd screenshot is where captions use the same priorities as correspondent icons).
The visual changes are surprisingly small (and its a city center! in less dense areas there will be hardly any differences). And they're are not bad IMO, in fact there are less lower importance (for this zl) icons.
And it'll make maintaining overlays priorities **a lot easier**.

Also if we're not happy with some changes then its very likely its better solved by adjusting captions' visibility.

@biodranik  @vng  WDYT?

Some museums captions replacing stuff around. And one park caption replacing a church:
![cap-ic_z14](https://github.com/organicmaps/organicmaps/assets/18434508/fd44c68d-60bf-410a-aa8c-7669a7b2306d)

![cap-ic_z16](https://github.com/organicmaps/organicmaps/assets/18434508/b846c280-b120-4005-b802-326cd53bf43c)

![cap-ic_z18](https://github.com/organicmaps/organicmaps/assets/18434508/fdb30328-2924-4a2e-9930-5107cdbb37a6)

pastk commented:
And here is a comparison of two "opposite" approaches:
1. Captions priorities == icon priorities (as in prev post).
2. Optional captions below other overlays always, i.e.
  - first go icons, pathtexts, shields and lone captions (e.g. a neighborhood or a building name).
  - below them are "optional" captions that have an icon

To me it looks good too! More icons and less texts makes the map cleaner (there are more housenumbers though).
And it makes priorities maintenance easier also (but still a bit more complicated than option 1 as we need to "sub-range" optional captions somehow).

![ic-over-caps_z14](https://github.com/organicmaps/organicmaps/assets/18434508/5fdc579a-90ad-4d2d-9582-0458772a576c)

![ic-over-caps_z16](https://github.com/organicmaps/organicmaps/assets/18434508/979d44a3-9e8a-44bc-8333-b1807cb0a8ba)

![ic-over-caps_z18](https://github.com/organicmaps/organicmaps/assets/18434508/6710eddd-35f6-48cb-a5f5-1ad46cd13478)

pastk commented:
Expanded priorities files syntax:
- added grouping of types with same priorities
- added automatic comments with visibility range and other drules info

```
highway-bus_stop                         icon       # z16- (also has caption z17-)
=== 1105

highway-pedestrian-bridge                pathtext   # z14- (also has line z13-, line::bridgeblack z14-, line::bridgewhite z13-)
highway-tertiary-tunnel                  shield     # z15- (also has line z11-, line(casing) z16-, pathtext z12-)
natural-strait                           caption    # z13- (also has line z13-)
waterway-river                           pathtext   # z11- (also has line z10-)
waterway-riverbank                       pathtext   # z11- (also has area z1-, line z10-)
=== 1100
```

pastk commented:
> 2\. Optional captions below other overlays always, i.e.
>     * first go icons, pathtexts, shields and lone captions (e.g. a neighborhood or a building name).
>     * below them are "optional" captions that have an icon

So I continued with that approach.

- removed explicit drule types from prio files, now its just type name with optional ::object_id
```
highway-bus_stop                                    # caption z17- and icon z16-
=== 1105

highway-pedestrian-bridge                           # pathtext z14- (also has line z13-, line::bridgeblack z14-, line::bridgewhite z13-)
highway-tertiary-tunnel                             # pathtext z12- and shield z15- (also has line z11-, line(casing) z16-)
natural-strait                                      # caption z13- (also has line z13-)
waterway-riverbank                                  # pathtext z11- (also has area z1-, line z10-)
=== 1100
```

Added basic visibility ranges validation, fixed following inconsistencies:
Main:
- add a missing icon to amenity-theater z15 (was a caption only)
- made amenity-community_center z15- (had no icon, only captions at z14-15)
- fixed natural-peak [!name] to be z15- (natural-peak [name] is z12-)

Vehicle:
- amenity-community_center as above
- fixed highway-service-driveway line from z14,16- to z15-

pastk commented:
Its ready to go Beta IMO.

Jean-BaptisteC commented:
Sports icons are missing here https://osm.org/go/0DF5Wx3M-?m=

pastk commented:
> > Sports icons are missing here https://osm.org/go/0DF5Wx3M-?m=

Fixed!

pastk commented:
@biodranik @vng 

I think its ready for merge and release.

Please merge https://github.com/organicmaps/kothic/pull/10 first.

After that I'll rebase, amend kothic update commits, remove temporary commits and squash what's possible (I will preserve the migration history).

pastk commented:
- amenity-townhall had only z19 caption and 0 priority, made it z16- with an icon and proper priority

- lowered most vending machines and reduced visibility from z17- to z18- (except parking tickets, transport tickets)
- removed most vending machines from vehicle style
- lowered recycling containers
- removed min-icon-distance from vending machines, amenity-recycling*, parcel lockers and some other minor POIs

pastk commented:
I've updated the first post with an overview of the changes in this PR.

pastk commented:
> After that I'll rebase, amend kothic update commits, remove temporary commits and squash what's possible (I will preserve the migration history).

Did all that and did a final overlays priorities re-spacing to make it beautiful and easy to edit.

Ready to merge.

vng commented on drape/overlay_handle.hpp:
```diff
@@ -165,7 +168,7 @@ class OverlayHandle
 
   void SetDisplayFlag(bool display) { m_displayFlag = display; }
-  bool GetDisplayFlag() const { return m_displayFlag; }
+  bool GetDisplayFlag() const { return true; } // @todo Remove all displayFlag logic as its not used anymore.
```
Ok, if we modify here, so worth removing m_displayFlag var, no?

pastk replied:
Yes, I will remove it in subsequent PRs together with associated logic.

vng replied:
Then I have doubts, what do we fix here? I see that functions:
```
handle->SetDisplayFlag(false);
handle->SetDisplayFlag(true);
```
are called, and the logic looks valid.

vng commented on drape_frontend/colored_symbol_shape.cpp:
```diff
@@ -109,7 +109,11 @@ void ColoredSymbolShape::Draw(ref_ptr<dp::GraphicsContext> context, ref_ptr<dp::
+  // (the displacement engine takes care of this).
+  // But passing it in will make shields out of [dp::kMinDepth, dp::kMaxDepth]
+  // not rendered at all - avoid this so that shields priority range is not restricted.
+  glsl::vec3 const position = glsl::vec3(pt, 0 /* depth */);
```
Well, here is a bit strange change, because ColoredSymbolShape may be not only a shield.
I didn't get the problem from the comment, could we make some other fix here?

pastk replied:
Its basically for all overlays - the depth value is not needed for OGL to render them because they never overlap (due to displacement).
But the depth value is limited by the [kMinDepth;kMaxDepth] range hence limits overlay priorities range unnecessary - e.g. during priorities migration I had it up to 200000 (and kMaxDepth is 25000).

pastk replied:
So if we don't pass the depth value here then its not getting limited by those bounds.

pastk replied:
I see its being used for user marks also, I'll double check.

pastk replied:
For user marks the depth seems to be set to kInvalidDepth always :)

But ok I'll remove this change for now just to play safe. Its not important now as priorities were compressed to fit into a range much smaller than [kMinDepth;kMaxDepth] anyway.

vng commented on indexer/displacement_manager.hpp:
```diff
@@ -174,11 +174,11 @@ class DisplacementManager
-      float const d = base::Clamp(depth, kMinDepth, kMaxDepth) - kMinDepth;
+      // @todo: make sure features are prioritised the same way as in the run-time displacer,
+      // see overlay_handle.cpp::CalculateOverlayPriority()
+      ASSERT(-10000 <= depth && depth < 10000, (depth));
```
dp::kMinDepth, dp::kMaxDepth ?

pastk replied:
nope its different (kMaxDepth == 25000)

pastk replied:
I'll add proper constants here later when refactoring this part.

vng commented:
All *_shape.hpp (XxxShape) classes are common drape API. They are used and will be used in other cases, so we can't just make _zero_ depth relying on overlays displacement.

GUI overlays are also using shapes. Like draw SymbolShape with depth = x and TextShape over it with depth = x+1.

If there are any problems with POIs, should create a shape with zero depth on the client side, but I still don't understand what is not working here.

pastk commented:
> All *_shape.hpp (XxxShape) classes are common drape API. They are used and will be used in other cases, so we can't just make _zero_ depth relying on overlays displacement.

Yeap I've removed this change to play safe.
PTAL!

vng commented on drape_frontend/colored_symbol_shape.cpp:
```diff
@@ -109,6 +109,11 @@ void ColoredSymbolShape::Draw(ref_ptr<dp::GraphicsContext> context, ref_ptr<dp::
+  // m_params.m_depth value to OGL doesn't make sense, but could lead to
+  // elements out of [dp::kMinDepth, dp::kMaxDepth] depth range being
+  // not rendered at all. Depth values are derived from priorities often
+  // hence it leads to unnecessary restriction of possible priorities range.
```
Let's put assert here, but these comments are not relevant. These are API classes.

pastk commented:
@vng Fixed the comments as discussed. PTAL!
